In [1]:
import re
import json
import pandas as pd
from glob import glob
from os.path import join,basename
from msp_tables import prepare_tables, log
# make things prettier
from rich import print
from rich.progress import track, Progress

log.setLevel("ERROR")

bidsdir = '/mnt/bulk-vega/paulkuntke/mspaths'



First we filter out all available patient-IDs as well as the dates of examination

In [2]:
mpis = [ basename(sub).replace('sub-','')  for sub in glob(join(bidsdir, 'sub-*'))]

Read all the session-dates - as defined in the BIDS-Structure

In [3]:
sessionfiles = glob(join(bidsdir, 'sub-*', 'sub-*_sessions.tsv'))

sessions_df = pd.DataFrame()

subject_pattern = re.compile(r"sub-(\d+)_")


for sesfile in track(sessionfiles):
    df = pd.read_csv(sesfile, delimiter='\t')
    df["subject"] = re.findall(subject_pattern, basename(sesfile))[0]
    sessions_df = pd.concat((sessions_df, df))

del(df)
del(sessionfiles)

sessions_df.acq_time = pd.to_datetime(sessions_df.acq_time)



Output()

In [4]:
sessions_df.drop_duplicates()

session_id   acq_time    subject
0     ses-001 2019-05-03  100052186
1     ses-002 2020-02-22  100052186
0     ses-001 2017-02-09  100006427
1     ses-002 2017-06-08  100006427
0     ses-001 2020-11-08  500002712
..        ...        ...        ...
1     ses-002 2017-11-24  100003594
2     ses-003 2019-12-29  100003594
0     ses-001 2019-06-17  500003493
1     ses-002 2020-07-10  500003493
2     ses-003 2021-07-11  500003493

[30750 rows x 3 columns]

Now- lets read all tables ( _all_ is defined in `column_names.json`)

In [6]:
with open("column_names.json", "r") as file:
    table_data = json.load(file)
prepared_tables = prepare_tables('888MS001 & 888MS002/Data Tables/', bidsdir, table_data)
prepared_tables_hc = prepare_tables('888MS005/Data Tables/', bidsdir, table_data)

/home/paulkuntke/mspaths/msp_tables.py:74: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  f_df = pd.read_csv(f, encoding="cp1252") # Some tables contain Chars that are not readable with default utf-8 codepage
/home/paulkuntke/mspaths/msp_tables.py:74: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  f_df = pd.read_csv(f, encoding="cp1252") # Some tables contain Chars that are not readable with default utf-8 codepage
/home/paulkuntke/mspaths/msp_tables.py:74: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  f_df = pd.read_csv(f, encoding="cp1252") # Some tables contain Chars that are not readable with default utf-8 codepage
/home/paulkuntke/mspaths/msp_tables.py:74: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  f_df = pd.read_csv(f, encoding="cp1252") # Some tables contain 

In [7]:
prepared_tables["MSPT Sociodemographics"]

mpi   site     sex                             file  \
1       100000012  275.0  female  mspt_sociodemographics_v003.csv   
4       100000017  275.0  female  mspt_sociodemographics_v003.csv   
5       100000021  275.0    male  mspt_sociodemographics_v003.csv   
6       100000023  275.0    male  mspt_sociodemographics_v003.csv   
7       100000026  275.0  female  mspt_sociodemographics_v003.csv   
...           ...    ...     ...                              ...   
201745  500003114  326.0  female  mspt_sociodemographics_v007.csv   
201746  500003116  326.0  female  mspt_sociodemographics_v007.csv   
201748  500003118  326.0    male  mspt_sociodemographics_v007.csv   
201749  500003119  396.0  female  mspt_sociodemographics_v007.csv   
201750  500003120  326.0  female  mspt_sociodemographics_v007.csv   

                             race  encounter_date               ethnicity  \
1                           white    1.461329e+09  not_hispanic_or_latino   
4                           white    1.461402e+09  not_hispanic_or_latino   
5                           white    1.461487e+09  not_hispanic_or_latino   
6       black_or_african_american    1.453642e+09  not_hispanic_or_latino   
7                           white    1.462006e+09  not_hispanic_or_latino   
...                           ...             ...                     ...   
201745             western_europe    1.540558e+09                     NaN   
201746             western_europe    1.541765e+09                     NaN   
201748             western_europe    1.541756e+09                     NaN   
201749             western_europe    1.527758e+09                     NaN   
201750             western_europe    1.542014e+09                     NaN   

        age_at_first_symptom  age_at_diagnosis  
1                       40.0              40.0  
4                       33.0              33.0  
5                       30.0              30.0  
6                       35.0              38.0  
7                       32.0              32.0  
...                      ...               ...  
201745                  31.0               NaN  
201746                  50.0              60.0  
201748                  40.0              49.0  
201749                  42.0              42.0  
201750                   NaN               NaN  

[131675 rows x 9 columns]

In [8]:
for k, df in prepared_tables.items():
    print(k)
    display(df)

MSPT Sociodemographics

mpi   site     sex                             file  \
1       100000012  275.0  female  mspt_sociodemographics_v003.csv   
4       100000017  275.0  female  mspt_sociodemographics_v003.csv   
5       100000021  275.0    male  mspt_sociodemographics_v003.csv   
6       100000023  275.0    male  mspt_sociodemographics_v003.csv   
7       100000026  275.0  female  mspt_sociodemographics_v003.csv   
...           ...    ...     ...                              ...   
201745  500003114  326.0  female  mspt_sociodemographics_v007.csv   
201746  500003116  326.0  female  mspt_sociodemographics_v007.csv   
201748  500003118  326.0    male  mspt_sociodemographics_v007.csv   
201749  500003119  396.0  female  mspt_sociodemographics_v007.csv   
201750  500003120  326.0  female  mspt_sociodemographics_v007.csv   

                             race  encounter_date               ethnicity  \
1                           white    1.461329e+09  not_hispanic_or_latino   
4                           white    1.461402e+09  not_hispanic_or_latino   
5                           white    1.461487e+09  not_hispanic_or_latino   
6       black_or_african_american    1.453642e+09  not_hispanic_or_latino   
7                           white    1.462006e+09  not_hispanic_or_latino   
...                           ...             ...                     ...   
201745             western_europe    1.540558e+09                     NaN   
201746             western_europe    1.541765e+09                     NaN   
201748             western_europe    1.541756e+09                     NaN   
201749             western_europe    1.527758e+09                     NaN   
201750             western_europe    1.542014e+09                     NaN   

        age_at_first_symptom  age_at_diagnosis  
1                       40.0              40.0  
4                       33.0              33.0  
5                       30.0              30.0  
6                       35.0              38.0  
7                       32.0              32.0  
...                      ...               ...  
201745                  31.0               NaN  
201746                  50.0              60.0  
201748                  40.0              49.0  
201749                  42.0              42.0  
201750                   NaN               NaN  

[131675 rows x 9 columns]

MSPT Patientcharacteristics

mpi   site             pdds_cat  pdds_scr  \
2       100000010  275.0  moderate_disability       2.0   
3       100000010  275.0      gait_disability       3.0   
4       100000010  275.0      mild_disability       1.0   
5       100000010  275.0      mild_disability       1.0   
6       100000010  275.0  moderate_disability       2.0   
...           ...    ...                  ...       ...   
725360  500013374  326.0               normal       0.0   
725410  500013406  326.0               normal       0.0   
725438  500013420  326.0               normal       0.0   
725785  500014006  326.0      mild_disability       1.0   
725992  500014219  326.0               normal       0.0   

                              mstype relapses              rel_stat  \
2             Relapsing Remitting MS        1                   NaN   
3                                NaN        1                   NaN   
4                                NaN        0                   NaN   
5                                NaN        0                   NaN   
6                                NaN        0                   NaN   
...                              ...      ...                   ...   
725360  Clinically Isolated Syndrome        0  domestic_partnership   
725410  Clinically Isolated Syndrome        0         never_married   
725438        Relapsing Remitting MS        0               married   
725785        Relapsing Remitting MS        0               married   
725992  Clinically Isolated Syndrome        0         never_married   

                                         file  encounter_date  \
2       mspt_patient characteristics_v010.csv    1.461320e+09   
3       mspt_patient characteristics_v010.csv    1.471075e+09   
4       mspt_patient characteristics_v010.csv    1.484397e+09   
5       mspt_patient characteristics_v010.csv    1.498912e+09   
6       mspt_patient characteristics_v010.csv    1.524225e+09   
...                                       ...             ...   
725360  mspt_patient_characteristics_v016.csv    1.631958e+09   
725410  mspt_patient_characteristics_v016.csv    1.633686e+09   
725438  mspt_patient_characteristics_v016.csv    1.633779e+09   
725785  mspt_patient_characteristics_v016.csv    1.642941e+09   
725992  mspt_patient_characteristics_v016.csv    1.651992e+09   

                     insurance employment_status current_living_situation  \
2                     medicaid          disabled  home_without_assistance   
3                     medicare          disabled     home_with_assistance   
4                     medicaid          disabled  home_without_assistance   
5                     medicare          disabled     home_with_assistance   
6                     medicaid          disabled  home_without_assistance   
...                        ...               ...                      ...   
725360  free_social_healthcare         full_time  home_without_assistance   
725410  free_social_healthcare         full_time  home_without_assistance   
725438  free_social_healthcare         part_time  home_without_assistance   
725785  free_social_healthcare         full_time  home_without_assistance   
725992  free_social_healthcare         part_time  home_without_assistance   

       number_of_relapses walking_aid walking_aid_one_hand  years_of_education  
2                     NaN         NaN                  NaN                12.0  
3                     NaN         NaN                  NaN                 NaN  
4                     NaN         NaN                  NaN                 NaN  
5                     NaN         NaN                  NaN                 NaN  
6                     NaN         NaN                  NaN                 NaN  
...                   ...         ...                  ...                 ...  
725360                NaN         NaN                  NaN                12.0  
725410                NaN         NaN                  NaN                16.0  
725438           

EMR Sociodemographics

mpi   site   age                            file  \
1       100000010  275.0  43.0  emr_sociodemographics_v012.csv   
2       100000011  275.0  63.0  emr_sociodemographics_v012.csv   
3       100000012  275.0  51.0  emr_sociodemographics_v012.csv   
6       100000016  275.0  48.0  emr_sociodemographics_v012.csv   
7       100000017  275.0  43.0  emr_sociodemographics_v012.csv   
...           ...    ...   ...                             ...   
223259  500013374  326.0  48.0  emr_sociodemographics_v016.csv   
223267  500013406  326.0  52.0  emr_sociodemographics_v016.csv   
223268  500013420  326.0  45.0  emr_sociodemographics_v016.csv   
223283  500014006  326.0  50.0  emr_sociodemographics_v016.csv   
223309  500014219  326.0  37.0  emr_sociodemographics_v016.csv   

        effective_date     sex   race               ethnicity  
1         1.583474e+09  Female  Black  Not Hispanic or Latino  
2         1.581661e+09    Male  White  Not Hispanic or Latino  
3         1.581575e+09  Female  White  Not Hispanic or Latino  
6         1.582265e+09  Female  White  Not Hispanic or Latino  
7         1.581661e+09  Female  White  Not Hispanic or Latino  
...                ...     ...    ...                     ...  
223259    1.674204e+09  Female    NaN                     NaN  
223267    1.675124e+09    Male    NaN                     NaN  
223268    1.675133e+09  Female    NaN                     NaN  
223283    1.676083e+09  Female    NaN                     NaN  
223309    1.676083e+09  Female    NaN                     NaN  

[142290 rows x 8 columns]

EMR Comorbidities

mpi   site cardio diabetes pain copd fatigue ibd   rc osteopor  \
1       100000010  275.0    Yes       No  Yes   No     Yes  No  Yes       No   
2       100000011  275.0    Yes      Yes  Yes   No      No  No   No       No   
3       100000012  275.0    Yes       No   No   No      No  No  Yes       No   
6       100000016  275.0     No       No  Yes   No     Yes  No   No       No   
7       100000017  275.0    Yes       No  Yes   No      No  No  Yes       No   
...           ...    ...    ...      ...  ...  ...     ...  ..  ...      ...   
147417  500003190  346.0     No       No   No   No      No  No   No       No   
147420  500003220  326.0     No       No   No   No      No  No   No       No   
147433  500003299  346.0     No       No   No   No      No  No   No       No   
147455  500003393  346.0     No       No   No   No      No  No   No       No   
147456  500003414  346.0     No       No   No   No      No  No   No       No   

       thyroid dyslipid depressn                        file  
1           No       No      Yes  emr_comorbidities_v010.csv  
2           No      Yes      Yes  emr_comorbidities_v010.csv  
3           No       No      Yes  emr_comorbidities_v010.csv  
6           No       No      Yes  emr_comorbidities_v010.csv  
7           No       No      Yes  emr_comorbidities_v010.csv  
...        ...      ...      ...                         ...  
147417      No       No       No  emr_comorbidities_v013.csv  
147420      No       No       No  emr_comorbidities_v013.csv  
147433      No       No      Yes  emr_comorbidities_v013.csv  
147455      No       No       No  emr_comorbidities_v013.csv  
147456      No       No       No  emr_comorbidities_v013.csv  

[92958 rows x 14 columns]

EMR Immunization

mpi   site      vac_date  vac_code  \
0      100014397  215.0  1.618031e+09     207.0   
1      100014397  215.0  1.615612e+09     207.0   
2      100005570  225.0  1.616216e+09     208.0   
3      100005570  225.0  1.613797e+09     208.0   
4      100004660  275.0  1.616735e+09     208.0   
...          ...    ...           ...       ...   
16501  100012788  225.0  1.618722e+09     208.0   
16502  100004977  245.0  1.623992e+09     208.0   
16504  100005338  225.0  1.616303e+09     208.0   
16505  100003068  275.0  1.612501e+09     207.0   
16506  100000764  275.0  1.615007e+09     207.0   

                                                vac_name  vac_dose vac_unit  \
0      SARS-COV-2 (COVID-19) vaccine, mRNA, spike pro...       0.5       mL   
1      SARS-COV-2 (COVID-19) vaccine, mRNA, spike pro...       0.5       mL   
2      SARS-COV-2 (COVID-19) vaccine, mRNA, spike pro...       0.3       mL   
3      SARS-COV-2 (COVID-19) vaccine, mRNA, spike pro...       0.3       mL   
4      SARS-COV-2 (COVID-19) vaccine, mRNA, spike pro...       0.3       mL   
...                                                  ...       ...      ...   
16501  SARS-COV-2 (COVID-19) vaccine, mRNA, spike pro...       0.3       mL   
16502  SARS-COV-2 (COVID-19) vaccine, mRNA, spike pro...       0.3       mL   
16504  SARS-COV-2 (COVID-19) vaccine, mRNA, spike pro...       0.3       mL   
16505  SARS-COV-2 (COVID-19) vaccine, mRNA, spike pro...       NaN      NaN   
16506  SARS-COV-2 (COVID-19) vaccine, mRNA, spike pro...       NaN      NaN   

                vac_mfr                       file  
0      Moderna US, Inc.  emr_immunization_v014.csv  
1      Moderna US, Inc.  emr_immunization_v014.csv  
2           Pfizer, Inc  emr_immunization_v014.csv  
3           Pfizer, Inc  emr_immunization_v014.csv  
4           Pfizer, Inc  emr_immunization_v014.csv  
...                 ...                        ...  
16501       Pfizer, Inc  emr_immunization_v017.csv  
16502       Pfizer, Inc  emr_immunization_v017.csv  
16504       Pfizer, Inc  emr_immunization_v017.csv  
16505  Moderna US, Inc.  emr_immunization_v017.csv  
16506  Moderna US, Inc.  emr_immunization_v017.csv  

[11688 rows x 9 columns]

EMR Socialhistory

mpi   site    smk_strt     smk_end      smk_code  \
0       100057522  225.0  1609477200  1609477200  2.669190e+08   
2       100007824  295.0  1565413200  1565413200  8.517006e+06   
3       100006602  245.0         NaN   923202000  7.717600e+07   
4       100099014  235.0  1673586000  1673586000  8.517006e+06   
6       100007669  295.0  1658466000  1658466000  8.517006e+06   
...           ...    ...         ...         ...           ...   
444012  100020234  285.0  1521867600         NaN  4.280410e+14   
444013  100009100  245.0  1554526800  1554526800  8.517006e+06   
444014  100018812  295.0  1532581200  1532581200  8.517006e+06   
444016  100007526  245.0  1618635600  1618635600  2.669190e+08   
444017  100007037  295.0         NaN         NaN  2.669190e+08   

                              smk_stat                         file  \
0                         Never smoker  emr_social_history_v017.csv   
2                        Former smoker  emr_social_history_v017.csv   
3       Smoker, current status unknown  emr_social_history_v017.csv   
4                        Former smoker  emr_social_history_v017.csv   
6                        Former smoker  emr_social_history_v017.csv   
...                                ...                          ...   
444012         Current some day smoker  emr_social history_v016.csv   
444013                   Former smoker  emr_social history_v016.csv   
444014                   Former smoker  emr_social history_v016.csv   
444016                    Never smoker  emr_social history_v016.csv   
444017                    Never smoker  emr_social history_v016.csv   

        effective_from_date_time  effective_to_date_time  snomed_code  \
0                            NaN                     NaN          NaN   
2                            NaN                     NaN          NaN   
3                            NaN                     NaN          NaN   
4                            NaN                     NaN          NaN   
6                            NaN                     NaN          NaN   
...                          ...                     ...          ...   
444012                       NaN                     NaN          NaN   
444013                       NaN                     NaN          NaN   
444014                       NaN                     NaN          NaN   
444016                       NaN                     NaN          NaN   
444017                       NaN                     NaN          NaN   

       smoking_status  
0                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
6                 NaN  
...               ...  
444012            NaN  
444013            NaN  
444014            NaN  
444016            NaN  
444017            NaN  

[307259 rows x 11 columns]

Social History

Empty DataFrame
Columns: [mpi]
Index: []

MSPT Neurological

mpi   site       encdate    mdt_avg     mdt_lh     mdt_rh  \
2       100000010  275.0  1.461320e+09        NaN        NaN        NaN   
3       100000010  275.0  1.471075e+09  37.336947  44.817707  29.856187   
4       100000010  275.0  1.484397e+09  41.902154  47.602544  36.201764   
5       100000010  275.0  1.498912e+09  44.409044  50.702549  38.115538   
6       100000010  275.0  1.524225e+09        NaN        NaN  33.714579   
...           ...    ...           ...        ...        ...        ...   
726124  100004527  245.0  1.487849e+09  23.564072  23.601758  23.526387   
726125  100004459  245.0  1.488187e+09  28.496457  33.460719  23.532195   
726126  100006715  245.0  1.488445e+09  29.826736  27.379463  32.274010   
726127  100007872  245.0  1.489239e+09  23.788993  25.353196  22.224790   
726128  100020567  235.0  1.501858e+09  23.255160  25.410542  21.099778   

       mdt_dh mdt_can mdt_canr    wst_avg  ... mdt_cancel mdt_cancel_reason  \
2         NaN     Yes   unable        NaN  ...        NaN               NaN   
3       right      No      NaN  12.324928  ...        NaN               NaN   
4       right      No      NaN   7.157859  ...        NaN               NaN   
5       right      No      NaN  10.932643  ...        NaN               NaN   
6       right      No      NaN  11.925833  ...        NaN               NaN   
...       ...     ...      ...        ...  ...        ...               ...   
726124  right      No      NaN   7.355662  ...        NaN               NaN   
726125  right      No      NaN   8.348110  ...        NaN               NaN   
726126  right      No      NaN   8.344832  ...        NaN               NaN   
726127  right      No      NaN   5.564342  ...        NaN               NaN   
726128  right      No      NaN  12.324767  ...        NaN               NaN   

       wst_afo_choice wst_walking_aid  wst_cancel  wst_cancel_reason  \
2                 NaN             NaN         NaN                NaN   
3                 NaN             NaN         NaN                NaN   
4                 NaN             NaN         NaN                NaN   
5                 NaN             NaN         NaN                NaN   
6                 NaN             NaN         NaN                NaN   
...               ...             ...         ...                ...   
726124            NaN             NaN         NaN                NaN   
726125            NaN             NaN         NaN                NaN   
726126            NaN             NaN         NaN                NaN   
726127            NaN             NaN         NaN                NaN   
726128            NaN             NaN         NaN                NaN   

       cst_cancel cst_cancel_reason  pst_cancel pst_cancel_reason  
2             NaN               NaN         NaN               NaN  
3             NaN               NaN         NaN               NaN  
4             NaN               NaN         NaN               NaN  
5             NaN               NaN         NaN               NaN  
6             NaN               NaN         NaN               NaN  
...           ...               ...         ...               ...  
726124        NaN               NaN         NaN               NaN  
726125        NaN               NaN         NaN               NaN  
726126        NaN               NaN         NaN               NaN  
726127        NaN               NaN         NaN               NaN  
726128        NaN               NaN         NaN               NaN  

[539740 rows x 43 columns]

EMR Medications

mpi   site      med_strt       med_end med_code  \
7        100000010  275.0  1517547600.0  1.518858e+09   106346   
8        100000010  275.0  1479704400.0  1.482556e+09  1085636   
9        100000010  275.0  1482555600.0  1.488172e+09  1085636   
10       100000010  275.0  1488430800.0  1.498712e+09  1085636   
11       100000010  275.0  1498712400.0  1.504328e+09  1085636   
...            ...    ...           ...           ...      ...   
6869536  500001835  346.0  1438902000.0           NaN  1373484   
6869537  500001835  346.0           NaN           NaN    25480   
6869538  500001835  346.0           NaN           NaN    E0143   
6869546  500001849  346.0  1141686000.0           NaN   135779   
6869547  500001849  346.0           NaN           NaN    E0100   

                                                  med_name  med_dose  \
7                    MUPIROCIN 0.02 MG/MG TOPICAL OINTMENT       1.0   
8        TRIAMCINOLONE ACETONIDE 0.001 MG/MG TOPICAL OI...       NaN   
9        TRIAMCINOLONE ACETONIDE 0.001 MG/MG TOPICAL OI...       NaN   
10       TRIAMCINOLONE ACETONIDE 0.001 MG/MG TOPICAL OI...       NaN   
11       TRIAMCINOLONE ACETONIDE 0.001 MG/MG TOPICAL OI...       NaN   
...                                                    ...       ...   
6869536                                          TECFIDERA     480.0   
6869537                                         GABAPENTIN     300.0   
6869538  WALKER, FOLDING, WHEELED, ADJUSTABLE OR FIXED ...       NaN   
6869546                                           COPAXONE       NaN   
6869547  CANE, INCLUDES CANES OF ALL MATERIALS, ADJUSTA...       NaN   

              med_unit                      file  
7        {APPLICATION}  emr_medications_v006.csv  
8                  NaN  emr_medications_v006.csv  
9                  NaN  emr_medications_v006.csv  
10                 NaN  emr_medications_v006.csv  
11                 NaN  emr_medications_v006.csv  
...                ...                       ...  
6869536             MG  emr_medications_v005.csv  
6869537             MG  emr_medications_v005.csv  
6869538            NaN  emr_medications_v005.csv  
6869546            NaN  emr_medications_v005.csv  
6869547            NaN  emr_medications_v005.csv  

[4818283 rows x 9 columns]

MRI Quantitative Metrics

mpi   site      sty_date      pri_date run_by          prog_ver  \
0       100000010  275.0  1447977600.0           NaN   BIIB  MSPiev2.0.0rc.p1   
1       100000010  275.0  1461283200.0  1447977600.0   BIIB  MSPiev2.0.0rc.p1   
2       100000010  275.0  1469318400.0  1461283200.0   BIIB  MSPiev2.0.0rc.p1   
3       100000010  275.0  1484352000.0  1469318400.0   BIIB  MSPiev2.0.0rc.p1   
4       100000010  275.0  1496016000.0  1484352000.0   BIIB  MSPiev2.0.0rc.p1   
...           ...    ...           ...           ...    ...               ...   
200008  100030710  275.0    1531440000  1500163200.0   BIIB      MSPiev0.1.0a   
200009  100031255  275.0    1499990400           NaN   BIIB      MSPiev0.1.0a   
200010  100031255  275.0    1527292800  1499990400.0   BIIB      MSPiev0.1.0a   
200011  100031762  275.0    1501372800           NaN   BIIB      MSPiev0.1.0a   
200012  100031762  275.0    1531612800  1501372800.0   BIIB      MSPiev0.1.0a   

        rad_acpt       bpf  bpf_chg  t2lesvol  ...  scanner      dgmf  \
0            NaN  0.795460      NaN   21.9116  ...    Skyra  0.023679   
1            NaN  0.798690  0.40605   20.4064  ...    Skyra  0.023464   
2            NaN  0.795880 -0.35183   20.6856  ...    Skyra  0.023754   
3            NaN  0.801200  0.66844   20.9312  ...    Skyra  0.023400   
4            NaN  0.805685  0.55979   20.0634  ...    Skyra  0.023466   
...          ...       ...      ...       ...  ...      ...       ...   
200008       NaN  0.888000  0.00600    2.1750  ...      NaN       NaN   
200009       NaN  0.826000      NaN    9.5660  ...      NaN       NaN   
200010       NaN  0.826000  0.04200   10.6090  ...      NaN       NaN   
200011       NaN  0.870000      NaN    3.1430  ...      NaN       NaN   
200012       NaN  0.867000 -0.32000    3.6780  ...      NaN       NaN   

            cgmf     thalf  t2overbv  t1dcpp  flrdcpp  nt2lesgt  \
0       0.373632  0.008384  0.023337       N        N       NaN   
1       0.384924  0.008230  0.021645       N        N       1.0   
2       0.387206  0.008491  0.021805       N        N       0.0   
3       0.376446  0.008233  0.022028       N        N       1.0   
4       0.398746  0.008374  0.021117       N        N       0.0   
...          ...       ...       ...     ...      ...       ...   
200008       NaN       NaN       NaN     NaN      NaN       NaN   
200009       NaN       NaN       NaN     NaN      NaN       NaN   
200010       NaN       NaN       NaN     NaN      NaN       NaN   
200011       NaN       NaN       NaN     NaN      NaN       NaN   
200012       NaN       NaN       NaN     NaN      NaN       NaN   

                                     file  study_id  
0       mri_quantitative_metrics_v015.csv       NaN  
1       mri_quantitative_metrics_v015.csv       NaN  
2       mri_quantitative_metrics_v015.csv       NaN  
3       mri_quantitative_metrics_v015.csv       NaN  
4       mri_quantitative_metrics_v015.csv       NaN  
...                                   ...       ...  
200008  mri_quantitative_metrics_v007.csv       NaN  
200009  mri_quantitative_metrics_v007.csv       NaN  
200010  mri_quantitative_metrics_v007.csv       NaN  
200011  mri_quantitative_metrics_v007.csv       NaN  
200012  mri_quantitative_metrics_v007.csv       NaN  

[197732 rows x 33 columns]

Biobanking

mpi   site  substudy     smk_code                  smk_stat  \
1        100017491  235.0       2.0  266927001.0    Unknown if ever smoked   
2        100061813  295.0       2.0  266927001.0    Unknown if ever smoked   
3        100039876  295.0       2.0  266927001.0    Unknown if ever smoked   
5        100040242  295.0       2.0  266927001.0    Unknown if ever smoked   
6        500001428  396.0       2.0  266927001.0    Unknown if ever smoked   
...            ...    ...       ...          ...                       ...   
2647151  100016824  235.0       2.0  449868002.0  Current every day smoker   
2647152  100016824  235.0       2.0  449868002.0  Current every day smoker   
2647153  100016824  235.0       2.0  449868002.0  Current every day smoker   
2647154  100016824  235.0       2.0  449868002.0  Current every day smoker   
2647155  100016824  235.0       2.0  449868002.0  Current every day smoker   

             specdate         specimen          subtype      lastmeal  \
1        1.498157e+09  AP984771A01-005   DNA WB ALIQUOT           NaN   
2        1.570304e+09      AP850848D01  DNA WHOLE BLOOD           NaN   
3        1.547834e+09  AP909818C01-004   DNA WB ALIQUOT           NaN   
5        1.569602e+09  AP842989D01-003   DNA WB ALIQUOT           NaN   
6        1.568565e+09      EP176562E01  DNA WHOLE BLOOD           NaN   
...               ...              ...              ...           ...   
2647151  1.497219e+09      AP992141A22  SERUM-MSPATHS-2  1.497208e+09   
2647152  1.497219e+09      AP992141A24   SERUM-BIOGEN-3  1.497208e+09   
2647153  1.497219e+09      AP992141A26   SERUM-BIOGEN-4  1.497208e+09   
2647154  1.497219e+09      AP992141A28  SERUM-MSPATHS-3  1.497208e+09   
2647155  1.497219e+09      AP992141A30  SERUM-MSPATHS-4  1.497208e+09   

         visit_nm  disposit visit_rn                           file  site_id  
1        REDACTED       1.0                     biobanking_v015.csv      NaN  
2        REDACTED       5.0                     biobanking_v015.csv      NaN  
3        REDACTED       1.0                     biobanking_v015.csv      NaN  
5        REDACTED       1.0                     biobanking_v015.csv      NaN  
6        REDACTED       1.0                     biobanking_v015.csv      NaN  
...           ...       ...      ...                            ...      ...  
2647151  BASELINE       1.0           biobanking_v012_CORRECTED.csv      NaN  
2647152  BASELINE       1.0           biobanking_v012_CORRECTED.csv      NaN  
2647153  BASELINE       1.0           biobanking_v012_CORRECTED.csv      NaN  
2647154  BASELINE       1.0           biobanking_v012_CORRECTED.csv      NaN  
2647155  BASELINE       1.0           biobanking_v012_CORRECTED.csv      NaN  

[1856762 rows x 14 columns]

Combine patients and HCs

In [9]:
mri_df = pd.concat((prepared_tables["MRI Quantitative Metrics"], prepared_tables_hc["MRI Quantitative Metrics"]))

# Convert Date-Column to same format everywhere
# 1. Identify which entries are numeric (Unix timestamps)
is_timestamp = pd.to_numeric(mri_df['sty_date'], errors='coerce').notna()

# 2. Convert timestamps (assumed to be seconds) to datetime
mri_df.loc[is_timestamp, 'date'] = pd.to_datetime(
    mri_df.loc[is_timestamp, 'sty_date'].astype(float), unit='s'
)

# 3. Convert the remaining string dates to datetime
mri_df.loc[~is_timestamp, 'date'] = pd.to_datetime(mri_df.loc[~is_timestamp, 'sty_date'])

# 4. Ensure the column dtype is datetime64[ns]
mri_df['date'] = pd.to_datetime(mri_df['date'])

mri_df = mri_df.merge(sessions_df, left_on=['mpi', 'date'], right_on=['subject', 'acq_time'], how='right').dropna(subset=['session_id', 'subject']).drop_duplicates(subset=['date', 'mpi'])
mri_df.query('session_id == "ses-001"').to_csv('baseline_quantiative_metrics.csv') 

mri_df

mpi   site      sty_date      pri_date run_by          prog_ver  \
0       100052186  225.0  1556841600.0           NaN   BIIB  MSPiev2.0.0rc.p1   
6       100052186  225.0  1582329600.0  1556841600.0   BIIB       MSPiev2.0.0   
8       100006427  285.0  1486598400.0           NaN   BIIB  MSPiev2.0.0rc.p1   
19      100006427  285.0  1496880000.0  1486598400.0   BIIB  MSPiev2.0.0rc.p1   
28      500002712  326.0  1604793600.0           NaN   BIIB       MSPiev2.0.0   
...           ...    ...           ...           ...    ...               ...   
193759  100003594  275.0  1511481600.0  1477612800.0   BIIB  MSPiev2.0.0rc.p1   
193769  100003594  275.0  1577577600.0  1511481600.0   BIIB  MSPiev2.0.0rc.p1   
193776  500003493  396.0  1560729600.0           NaN   BIIB  MSPiev2.0.0rc.p1   
193782  500003493  396.0  1594339200.0  1560729600.0   BIIB       MSPiev2.0.0   
193785  500003493  396.0  1625961600.0  1594339200.0   BIIB       MSPiev2.0.0   

        rad_acpt       bpf  bpf_chg  t2lesvol  ...  t2overbv  t1dcpp  flrdcpp  \
0            NaN  0.844728      NaN   38.2606  ...  0.029835       N        N   
6            NaN  0.837414 -0.86593   33.4038  ...  0.026759       N        N   
8            NaN  0.797080      NaN   34.3370  ...  0.040321       N        N   
19           NaN  0.801253  0.52353   33.2679  ...  0.039202       N        N   
28           NaN  0.788259      NaN   29.6135  ...  0.028293       N        N   
...          ...       ...      ...       ...  ...       ...     ...      ...   
193759       NaN  0.856995 -0.18956   10.8606  ...  0.009806       N        N   
193769       NaN  0.854145 -0.33259   10.4541  ...  0.009512       N        N   
193776       NaN  0.873024      NaN    0.1865  ...  0.000172       Y        Y   
193782       NaN  0.869819 -0.36713    0.2813  ...  0.000257       N        N   
193785       NaN  0.877134  0.84096    0.2662  ...  0.000239       N        N   

        nt2lesgt                               file  study_id       date  \
0            NaN  mri_quantitative_metrics_v015.csv       NaN 2019-05-03   
6            8.0  mri_quantitative_metrics_v017.csv       NaN 2020-02-22   
8            NaN  mri_quantitative_metrics_v015.csv       NaN 2017-02-09   
19           0.0  mri_quantitative_metrics_v015.csv       NaN 2017-06-08   
28           NaN  mri_quantitative_metrics_v015.csv       NaN 2020-11-08   
...          ...                                ...       ...        ...   
193759       0.0  mri_quantitative_metrics_v015.csv       NaN 2017-11-24   
193769       0.0  mri_quantitative_metrics_v015.csv       NaN 2019-12-29   
193776       NaN  mri_quantitative_metrics_v015.csv       NaN 2019-06-17   
193782       0.0  mri_quantitative_metrics_v015.csv       NaN 2020-07-10   
193785       0.0  mri_quantitative_metrics_v015.csv       NaN 2021-07-11   

        session_id   acq_time    subject  
0          ses-001 2019-05-03  100052186  
6          ses-002 2020-02-22  100052186  
8          ses-001 2017-02-09  100006427  
19         ses-002 2017-06-08  100006427  
28         ses-001 2020-11-08  500002712  
...            ...        ...        ...  
193759     ses-002 2017-11-24  100003594  
193769     ses-003 2019-12-29  100003594  
193776     ses-001 2019-06-17  500003493  
193782     ses-002 2020-07-10  500003493  
193785     ses-003 2021-07-11  500003493  

[26782 rows x 37 columns]

Now match MRI-QUantiative values with our sessions. 

 => maybe we should reverse this. 
Currently we get 26782 sessions but have 30750 in the BIDS-Dataset => What happened here?
Should identify


## MSPT patient Characteristics

In This table we find essential things like MS-Type, PDDS-score, etc.



In [10]:
patchar_df = prepared_tables["MSPT Patientcharacteristics"]
patchar_df["date"] = pd.to_datetime(patchar_df.encounter_date, unit='s')
patchar_df.drop_duplicates(inplace=True)



def find_best_characteristics(mpi, date, progress, task):
    progress.update(task, advance=1)
    
    patient_df = patchar_df.query(f'mpi == "{mpi}"')

    patient_df["distance"] = abs(patient_df.date - date)
    patient_df.sort_values('distance', inplace=True)

    if len(patient_df) == 0:
        return  pd.Series([
        
            pd.NA, pd.NA, pd.NA
        ])

    
    best_result = patient_df.iloc[0,:]
    return pd.Series([
        best_result["date"],
        best_result["mstype"],
        best_result["pdds_scr"]
         ])



with Progress() as progress:
    task = progress.add_task("Matching Sessions", total=len(mri_df))
    mri_df[["characteristics_date", "mstype", "pdds_scr"]] = mri_df.apply(lambda x: find_best_characteristics(x.mpi, x.date, progress, task), axis=1 )
   

Output()

In [33]:
neurological_df = pd.concat((prepared_tables["MSPT Neurological"],prepared_tables_hc["MSPT Neurological"] ))

neurological_df["date"] = pd.to_datetime(neurological_df.encdate, unit='s')
neurological_df.drop_duplicates(inplace=True)



def find_best_characteristics(mpi, date, progress, task):
    progress.update(task, advance=1)
    
    patient_df = patchar_df.query(f'mpi == "{mpi}"')

    patient_df["distance"] = abs(patient_df.date - date)
    patient_df.sort_values('distance', inplace=True)

    if len(patient_df) == 0:
        return  pd.Series([
        
            pd.NA, pd.NA, pd.NA
        ])

    
    best_result = patient_df.iloc[0,:]
    return pd.Series([
        best_result["date"],
        best_result["mstype"],
        best_result["pdds_scr"]
         ])



# with Progress() as progress:
 #   task = progress.add_task("Matching Sessions", total=len(mri_df))
  #  mri_df[["characteristics_date", "mstype", "pdds_scr"]] = mri_df.apply(lambda x: find_best_characteristics(x.mpi, x.date, progress, task), axis=1 )
   


Now add Data from other Tables 

EMR Social History:
- smk_code
- smk_stat

    for HCs these are in Biobanking
- smk_code
 -smk_stat

MSPT patient characteristics
 - educ

mspt neurological assesments

## Add Smoking Status

In [30]:
biobanking_df = pd.concat((prepared_tables["Biobanking"], prepared_tables_hc["Biobanking"]))[["mpi", "smk_code", "smk_stat"]]
biobanking_df.drop_duplicates(inplace=True)


biobanking_df.groupby('mpi').agg('count')# .query('smk_code == 1')

smk_code  smk_stat
mpi                          
100000010         3         3
100000011         3         3
100000012         2         2
100000016         2         2
100000017         2         2
...             ...       ...
500012902         2         2
500012925         3         3
500012943         2         2
500012948         2         2
500013031         2         2

[6870 rows x 2 columns]

In [ ]:
biobanking_df.query('mpi == "10000002"')

mpi     smk_code                             smk_stat
2418     100000024    8517006.0                        Former smoker
4024     100000024  449868002.0             Current every day smoker
5756     100000024  266919005.0                         Never smoker
17582    100000024  266927001.0               Unknown if ever smoked
384472   100000024  266927001.0  Tobacco smoking consumption unknown
387072   100000024    8517006.0                            Ex-smoker
1861432  100000024    8517006.0                        Former Smoker

In [ ]:
mri_df.to_hdf('mri_data.hdf5', key='mri_df')
